### PR TITLE
[Exporter.Geneva] Add exemplar support for otlp protobuf format

### LINKED ISSUE #1685: Add support for exporting Exemplars when enabling OTLP protobuf encoding
Opening the issue to track progress on adding exemplars support when OTLP protobuf encoding is enabled via `PrivatePreviewEnableOtlpProtobufEncoding`

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Fixes #1685 

## Changes

Adds support for exporting exemplars when OTLP protobuf encoding is enabled for metrics. TODO: Updated Readme to remove issue references.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs [hunk 1]
```diff
@@ -444,14 +452,76 @@ private void WriteNumberDataPoint<T>(byte[] buffer, ref int cursor, int fieldNum
...
+#endif
 
         var metricPointStartPosition = this.metricPointTagAndLengthIndex;
 
         // Write numberdatapoint {Repeated field}
         ProtobufSerializerHelper.WriteTagAndLengthPrefix(buffer, ref metricPointStartPosition, cursor - this.metricPointValueIndex, fieldNumber, WireType.LEN);
     }
 
+#if EXPOSE_EXPERIMENTAL_FEATURES
+    private void SerializeExemplar<T>(byte[] buffer, ref int cursor, in Exemplar exemplar, T value, int fieldNumber)
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs:479]: Mark this method and `SerializeExemplarTags` static.

File: test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs [hunk 2]
```diff
@@ -69,27 +93,86 @@ public OtlpProtobufMetricExporterTests()
...
-    [InlineData("longcounter", -123L, null, true, true)]
-    [InlineData("longcounter", -123L, null, true, false)]
-    [InlineData("longcounter", -123L, null, false, true)]
-    [InlineData("longcounter", -123L, null, false, false)]
-    [InlineData("doublecounter", null, -123.45, true, true)]
-    [InlineData("doublecounter", null, -123.45, true, false)]
-    [InlineData("doublecounter", null, -123.45, false, true)]
-    [InlineData("doublecounter", null, -123.45, false, false)]
-    public void CounterSerializationSingleMetricPoint(string instrumentName, long? longValue, double? doubleValue, bool addPrepopulatedDimensions, bool addAccountAndNamespace)
+    [InlineData("longcounter", 123L, null, true, true, true, true)]
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs:102]: This is huge number of test cases for `InlineData` based approach. You might want to check if there's a better way to test these cases. Maybe using xUnit ClassData? https://andrewlock.net/creating-parameterised-tests-in-xunit-with-inlinedata-classdata-and-memberdata/

@CodeBlanch Any suggestions?

File: src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs (hunk 1 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Exporter.Geneva/Metrics/OtlpProtobuf/OtlpProtobufSerializer.cs:479]: I am planning to do a follow up and mark the methods appropriately (There are few other methods that needs change). Did not want to do it here as that requires moving things around.

File: test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs (hunk 2 above)
Comment (vishweshbankwar) [reply test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs:102]: Yeah, this probably needs some rewrite. I was considering to decouple some testing in new tests. It's the number of possible combinations that is making this big.

File: test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs [hunk 3]
```diff
@@ -187,6 +292,49 @@ public void CounterSerializationSingleMetricPoint(string instrumentName, long? l
...
+            }
+            else
+            {
+                Assert.Equal(doubleValue.Value, exemplar.AsDouble);
+            }
+
+            if (isTracingEnabled)
+            {
+                var spanIdBytes = new byte[8];
+                activity.SpanId.CopyTo(spanIdBytes);
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs:313]: Add assert statements to ensure that `exemplar.TraceId` and `exemplar.SpanId` are not `default` values.

File: test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs [hunk 4]
```diff
@@ -187,6 +292,49 @@ public void CounterSerializationSingleMetricPoint(string instrumentName, long? l
...
+
+                for (int i = 0; i < 8; i++)
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs:315]: nit: Use `Enumerable.SequenceEqual` to check if the two arrays have the same values.

File: test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs [hunk 5]
```diff
@@ -187,6 +292,49 @@ public void CounterSerializationSingleMetricPoint(string instrumentName, long? l
             Assert.Equal(doubleValue, dataPoint.AsDouble);
         }
 
+#if EXPOSE_EXPERIMENTAL_FEATURES
+        if (isExemplarsEnabled)
+        {
+            Assert.Single(dataPoint.Exemplars);
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs:298]: Add a check for exemplar timestamp as well.

File: test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs (hunk 3 above)
Comment (vishweshbankwar) [reply test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs:316]: SequenceEqual checks that already as it compares it with activity spanId and traceId.

File: test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs (hunk 2 above)
Comment (utpilla) [reply test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs:102]: It's probably okay to remove the combinations for account and namespace for exemplar related tests.

File: test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs [hunk 6]
```diff
@@ -187,6 +292,50 @@ public void CounterSerializationSingleMetricPoint(string instrumentName, long? l
...
+
+            var exemplar = dataPoint.Exemplars[0];
+
+            Assert.NotEqual(default, exemplar.TimeUnixNano);
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs:302]: We could have a stricter assertion by ensuring that `exemplar.TimeUnixNanon` lies within an acceptable range. Something like this:
```csharp
var start = DateTime.UtcNow; //convert to unix nanoseconds
counter.Add(longValue.Value, this.exemplarTagList);
var end = DateTime.UtcNow;
...
Assert.True(exemplar.TimeUnixNano> start && exemplar.TimeUnixNano < end);
```

Comment (utpilla) [reply test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs:302]: Or maybe compare it with the exact timestamp by inspecting the Exemplars in the MetricPoint.

**Review by utpilla — APPROVED** [review]
Left some suggestions for improving tests. Looks good overall!

File: test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs (hunk 6 above)
Comment (vishweshbankwar) [reply test/OpenTelemetry.Exporter.Geneva.Tests/OtlpProtobufMetricExporterTests.cs]: Added

**Pull request merged**